## Rubric

Instructions: DELETE this cell before you submit via a `git push` to your repo before deadline. This cell is for your reference only and is not needed in your report. 

 Scoring: Out of 10 points

- Each Developing  => -2 pts
- Each Unsatisfactory/Missing => -4 pts
  - until the score is 0

If students address the detailed feedback in a future checkpoint they will earn these points back






|                                  | **Unsatisfactory**                                                                                                                                                                                                                                                                                                                        | **Developing**                                                                                                                                                                                                       | **Proficient**                                                                                                                                                                                            | **Excellent**                                                                                                                                                                            |
|----------------------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| **EDA relevance**                | EDA is mostly neither relevant to the question nor helpful in figuring out how to address the question. Or the EDA does address the question, but many obviously relevant variables / analyses / figures were not included. | EDA is partly irrelevant/unhelpful. EDA missed one or two obvioulsy relevant analysis (distributions of single variables or relationships between variables) | EDA includes the obviously relevant / helpful variables in addressing the question.                                                              | Thorough EDA fully explored the dataset                                                                                                                 |
| **EDA analysis and description** | Many of the analyses are poor choices (e.g., using means instead of medians for obviously skewed data), or are poorly described in the text, or do not aid understanding the data                                                                                                                                                     | Some of the analyses are poor choices, or are poorly described in the text, or do not aid understanding the data                                                                                                 | All analyses are correct choices. Only one or two have minor issues in the text descriptions supporting them. Mostly they fit well with other elements of the EDA and support understanding the data  | All analyses are correct choices with clear text descriptions supporting them. The figures fit well with the other elements of the EDA, producing a clear understanding of the data. |
| **EDA figures**                  | Many of the figures are poor plot choices (e.g., using a bar plot to represent a time series where it would be better to use a line plot) or have poor aesthetics (including colormap, data point shape/color, axis labels, titles, annotations, text legibility) or do not aid understanding the data                                | Some of the figures are poor plot choices or have poor aesthetics. Some figures do not aid understanding the data                                                                                                | All figures are correct plot choices. Only one or two have minor questionable aesthetic choices. The figures mostly fit well with the other elements of the EDA and support understanding the data    | All figures are correct plot choices with beautiful aesthetics. The figures fit well with the other elements of the EDA, producing a clear understanding of the data.                |





# COGS 108 - EDA Checkpoint

## Authors

Instructions: REPLACE the contents of this cell with your team list and their contributions. Note that this will change over the course of the checkpoints

This is a modified [CRediT taxonomy of contributions](https://credit.niso.org). For each group member please list how they contributed to this project using these terms:
> Analysis, Background research, Conceptualization, Data curation, Experimental investigation, Methodology, Project administration, Software, Visualization, Writing – original draft, Writing – review & editing

Example team list and credits:
- Alice Anderson: Conceptualization, Data curation, Methodology, Writing - original draft
- Bob Barker:  Analysis, Software, Visualization
- Charlie Chang: Project administration, Software, Writing - review & editing
- Dani Delgado: Analysis, Background research, Visualization, Writing - original draft

# Research Question

Instructions: REPLACE the contents of this cell with your work, including any updates to recover points lost in your proposal feedback



## Background and Prior Work

Instructions: REPLACE the contents of this cell with your work, including any updates to recover points lost in your proposal feedback

# Hypothesis


Instructions: REPLACE the contents of this cell with your work, including any updates to recover points lost in your proposal feedback

## Data

### Data overview

Instructions: REPLACE the contents of this cell with your work, including any updates to recover points lost in your data checkpoint feedback


In [46]:
# Imports and Setup
import pandas as pd
import numpy as np
import os

RAW_DATA_DIR = 'data/00-raw/'
INT_DATA_DIR = 'data/01-interim/'
PROCESSED_DATA_DIR = 'data/02-processed/'

In [3]:
# Run this code every time when you're actively developing modules in .py files.  It's not needed if you aren't making modules
#
## this code is necessary for making sure that any modules we load are updated here 
## when their source code .py files are modified

%load_ext autoreload
%autoreload 2

In [53]:
# Setup code -- Run only once after cloning!!! 
#
# this code downloads the data from its source to the `data/00-raw/` directory
# if the data hasn't updated you don't need to do this again!

# if you don't already have these packages (you should!) uncomment this line
# %pip install requests tqdm

import sys
sys.path.append('./modules') # this tells python where to look for modules to import

import get_data # this is where we get the function we need to download data

# replace the urls and filenames in this list with your actual datafiles
# yes you can use Google drive share links or whatever
# format is a list of dictionaries; 
# each dict has keys of 
#   'url' where the resource is located
#   'filename' for the local filename where it will be stored 
datafiles = [
    {
        'url': 'https://archive.ics.uci.edu/static/public/697/predict+students+dropout+and+academic+success.zip',
        'filename': 'predict_students_dropout.zip'
    },
    { 
        'url': 'https://zenodo.org/records/17239943/files/dataset_2022_hash.zip?download=1', 
        'filename':'university_dropout_2022.zip'
    }, # Decompressed later using pd.read_csv
    { 
        'url': 'https://zenodo.org/records/17239943/files/dataset_2021_hash.zip?download=1', 
        'filename':'university_dropout_2021.zip'
    } , # Decompressed later using pd.read_csv
    { 
        'url': 'https://zenodo.org/records/17239943/files/dataset_2018_hash.zip?download=1', 
        'filename':'university_dropout_2018.zip'
    } # Decompressed later using pd.read_csv
]
get_data.get_raw(datafiles,destination_directory='data/00-raw/')

Overall Download Progress:   0%|          | 0/4 [00:00<?, ?it/s]
                                                             

Successfully downloaded: predict_students_dropout.zip



Overall Download Progress:  50%|█████     | 2/4 [00:23<00:23, 11.58s/it]                    

Successfully downloaded: university_dropout_2022.zip



Overall Download Progress:  75%|███████▌  | 3/4 [00:25<00:07,  7.91s/it]                     

Successfully downloaded: university_dropout_2021.zip



Overall Download Progress: 100%|██████████| 4/4 [00:28<00:00,  7.10s/it]                     

Successfully downloaded: university_dropout_2018.zip


### Dataset #1 

Instructions: REPLACE the contents of this cell and the one below with your work, including any updates to recover points lost in your data checkpoint feedback

In [4]:
## YOUR CODE TO LOAD/CLEAN/TIDY/WRANGLE THE DATA GOES HERE
## FEEL FREE TO ADD MULTIPLE CELLS PER SECTION 

In [5]:
## YOUR CODE TO LOAD/CLEAN/TIDY/WRANGLE THE DATA GOES HERE
## 3.A LOAD DATASET
data1 = pd.read_csv(
    f'{RAW_DATA_DIR}predict_students_dropout.zip',
    sep=';',
    compression='zip' # Webpage download default as .zip
)

In [6]:
## 3.B MAKE TIDY OR SHOW TIDY
data1.head(5)

,Marital status,Application mode,Application order,Course,Daytime/evening attendance\t,Previous qualification,Previous qualification (grade),Nacionality,Mother's qualification,Father's qualification,...,Curricular units 2nd sem (credited),Curricular units 2nd sem (enrolled),Curricular units 2nd sem (evaluations),Curricular units 2nd sem (approved),Curricular units 2nd sem (grade),Curricular units 2nd sem (without evaluations),Unemployment rate,Inflation rate,GDP,Target
0,1,17,5,171,1,1,122.0,1,19,12,...,0,0,0,0,0.000000,0,10.8,1.4,1.74,Dropout
1,1,15,1,9254,1,1,160.0,1,1,3,...,0,6,6,6,13.666667,0,13.9,-0.3,0.79,Graduate
2,1,1,5,9070,1,1,122.0,1,37,37,...,0,6,0,0,0.000000,0,10.8,1.4,1.74,Dropout
3,1,17,2,9773,1,1,122.0,1,38,37,...,0,6,10,5,12.400000,0,9.4,-0.8,-3.12,Graduate
4,2,39,1,8014,0,1,100.0,1,37,38,...,0,6,6,6,13.000000,0,13.9,-0.3,0.79,Graduate


In [7]:
## 3.C DEMONSTRATE SIZE OF DATASET
print("Dataset shape (rows, columns):", data1.shape)
data1.dtypes

Dataset shape (rows, columns): (4424, 37)


Marital status                                      int64
Application mode                                    int64
Application order                                   int64
Course                                              int64
Daytime/evening attendance\t                        int64
Previous qualification                              int64
Previous qualification (grade)                    float64
Nacionality                                         int64
Mother's qualification                              int64
Father's qualification                              int64
Mother's occupation                                 int64
Father's occupation                                 int64
Admission grade                                   float64
Displaced                                           int64
Educational special needs                           int64
Debtor                                              int64
Tuition fees up to date                             int64
Gender        

**This data set was already used for more formal projects and has already been cleaned of missing values.**

In [8]:
## 3.D FIND OUT HOW MUCH DATA IS MISSING AND WHERE
col_missing_count = data1.isnull().sum()
total_missing_count = col_missing_count.sum()
total_missing_count
print(f"There are {total_missing_count} missing values.")

There are 0 missing values.


In [9]:
## 3.E FIND AND FLAG ANY OUTLIERS OR SUS ENTRIES
numeric_cols = data1.select_dtypes(include = ['number']).columns
outliers = pd.DataFrame(index = data1.index)
## iterate accross all numeric columns and find outliers and store in outliers df
for col in numeric_cols:
    ##https://stackoverflow.com/questions/23228244/how-do-you-find-the-iqr-in-numpy
    q75, q25 = np.nanpercentile(data1[col], [75, 25])
    iqr = q75 - q25
    lower_bound = q25 - 1.5 * iqr
    upper_bound = q75 + 1.5 * iqr
    outliers[f'{col}_outlier'] = (data1[col] < lower_bound) | (data1[col] > upper_bound)

print('Number of outliers in each column')
print(outliers.sum())

Number of outliers in each column
Marital status_outlier                                     505
Application mode_outlier                                     0
Application order_outlier                                  541
Course_outlier                                             442
Daytime/evening attendance\t_outlier                       483
Previous qualification_outlier                             707
Previous qualification (grade)_outlier                     179
Nacionality_outlier                                        110
Mother's qualification_outlier                               0
Father's qualification_outlier                               0
Mother's occupation_outlier                                182
Father's occupation_outlier                                177
Admission grade_outlier                                     86
Displaced_outlier                                            0
Educational special needs_outlier                           51
Debtor_outlier       

**This data is already very clean, no missing values or ridiculous outliers, the only thing that we may have to change is the data types being used for certain variables**

In [10]:
## 3.F CLEAN THE DATA
if data1.isna().sum().sum() == 0:
    print('No missing values')
else: 
    print(f'There are {data1.isna().sum().sum()} missin values')

No missing values


In [11]:
## Next need to remove columns that we do not need
cols_to_drop = [
    'Marital status',  # Demographics
    'Application mode', # Application related
    'Application order', # Application related
    'Course', # Student "Major"
    'Daytime/evening attendance\t', # Unnecessary and no equivalent in data2_2022
    'Previous qualification', # Students highest level of education
    'Nacionality', # Demographics
    "Mother's qualification", "Father's qualification", "Mother's occupation", "Father's occupation", # Family Demographics
    'Displaced', 'Educational special needs', 'Debtor', 'Tuition fees up to date', 'Scholarship holder', # Unnecessary Student demographics
    'Age at enrollment', 'International', # Unnecessary Student demographics
    'Unemployment rate', 'Inflation rate', 'GDP', # Portugal Economic Data
]

data1 = data1.drop(columns = cols_to_drop, errors = 'ignore')

In [12]:
## Mapping 'Target' to numerical values
target_map = {
    'Dropout': 0,
    'Enrolled': 1,
    'Graduate': 2
}

data1['Target_Numeric'] = data1['Target'].map(target_map)
print(data1[['Target', 'Target_Numeric']].head(100))
## Target_numeric now shows the outcome as an integer

      Target  Target_Numeric
0    Dropout               0
1   Graduate               2
2    Dropout               0
3   Graduate               2
4   Graduate               2
..       ...             ...
95  Enrolled               1
96  Graduate               2
97  Graduate               2
98  Enrolled               1
99   Dropout               0

[100 rows x 2 columns]


In [13]:
## Renaming cols to something that makes more sense
name_map = {
    # 1st Sem
    'Curricular units 1st sem (enrolled)': 'Sem1_Enrolled', # Credits Enrolled 
    'Curricular units 1st sem (credited)': 'Sem1_Credited', # Transfer credits
    'Curricular units 1st sem (approved)': 'Sem1_Passed', # Classes passed
    'Curricular units 1st sem (grade)': 'Sem1_Gpa',
    'Curricular units 1st sem (evaluations)': 'Sem1_Evaluations', # Classes where student took an exam
    'Curricular units 1st sem (without evaluations)': 'Sem1_Skipped', # Classes stopped attending/skipped

    # 2nd Sem
    'Curricular units 2nd sem (enrolled)': 'Sem2_Enrolled',
    'Curricular units 2nd sem (credited)': 'Sem2_Credited',
    'Curricular units 2nd sem (approved)': 'Sem2_Passed',
    'Curricular units 2nd sem (grade)': 'Sem2_Gpa',
    'Curricular units 2nd sem (evaluations)': 'Sem2_Evaluations',
    'Curricular units 2nd sem (without evaluations)': 'Sem2_Skipped'
}

data1 = data1.rename(columns = name_map)

In [14]:
## 3.G MOVE TO PROCESSED
data1_clean = data1.copy()
processed_path = os.path.join(PROCESSED_DATA_DIR, 'predict_students_dropout_clean.csv')
data1_clean.to_csv(processed_path, index=False, sep = ';')

In [15]:
## 4. SUMMARY STATISTICS
data1_clean.describe()

,Previous qualification (grade),Admission grade,Gender,Sem1_Credited,Sem1_Enrolled,Sem1_Evaluations,Sem1_Passed,Sem1_Gpa,Sem1_Skipped,Sem2_Credited,Sem2_Enrolled,Sem2_Evaluations,Sem2_Passed,Sem2_Gpa,Sem2_Skipped,Target_Numeric
count,4424.000000,4424.000000,4424.000000,4424.000000,4424.000000,4424.000000,4424.000000,4424.000000,4424.000000,4424.000000,4424.000000,4424.000000,4424.000000,4424.000000,4424.000000,4424.000000
mean,132.613314,126.978119,0.351718,0.709991,6.270570,8.299051,4.706600,10.640822,0.137658,0.541817,6.232143,8.063291,4.435805,10.230206,0.150316,1.178119
std,13.188332,14.482001,0.477560,2.360507,2.480178,4.179106,3.094238,4.843663,0.690880,1.918546,2.195951,3.947951,3.014764,5.210808,0.753774,0.888243
min,95.000000,95.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
25%,125.000000,117.900000,0.000000,0.000000,5.000000,6.000000,3.000000,11.000000,0.000000,0.000000,5.000000,6.000000,2.000000,10.750000,0.000000,0.000000
50%,133.100000,126.100000,0.000000,0.000000,6.000000,8.000000,5.000000,12.285714,0.000000,0.000000,6.000000,8.000000,5.000000,12.200000,0.000000,1.000000
75%,140.000000,134.800000,1.000000,0.000000,7.000000,10.000000,6.000000,13.400000,0.000000,0.000000,7.000000,10.000000,6.000000,13.333333,0.000000,2.000000
max,190.000000,190.000000,1.000000,20.000000,26.000000,45.000000,26.000000,18.875000,12.000000,19.000000,23.000000,33.000000,20.000000,18.571429,12.000000,2.000000


### University Student Dropout Dataset

The University Student Dropout dataset is organized as yearly CSV files named dataset_{year}.csv, with each row corresponding to a student-course enrollment for that academic year. Each file integrates data from four sources: students, programs, courses, and digital logs, and also groups variables into six thematic categories: context, admission pathways, socio-economic and demographic background, academic data, digital logs, and Wi-Fi access. Contextual attributes include anonymized identifiers for students, courses, academic programs, and campuses, as well as the academic year and group IDs, capturing where and how each student is enrolled. Admission pathway variables describe how the student entered the university, including year of enrollment, type of admission, entry exam grades (scaled to 10 or 14), and program selection preference. Socioeconomic and demographic variables capture parental education, student dedication to studies, and whether the student had to move provinces to attend university, providing insight into economic or social challenges that might affect retention.

Academic data is the most detailed category, including grades, credits enrolled and earned across multiple years, semester performance, adjustments for credit recognition, internships, activities, and overall progress toward degree completion. Metrics like cumulative GPA, credits passed per semester, and credit completion rates across previous years allow for longitudinal assessment of academic success and dropout risk. Digital logs track Learning Management System (LMS) site engagement monthly, including number of visits, events, assignment and test submissions, total minutes spent online, and usage of course resources. For 2021 and 2022, Wi-Fi access records provide an additional proxy for on-campus presence, recording the number of days each student accessed the university network per month. All variables are anonymized using hash codes, and numerical metrics such as grades are scaled (e.g., 0–10 or 0–14 for entry exams), while credit counts are in academic credit units. LMS and Wi-Fi activity metrics are counts of actions, logins, or days.

While these metrics provide valuable insights, several concerns about the dataset should be noted. It is drawn from a single Spanish technological university, which limits generalizability to other fields or institutions, particularly in humanities or social sciences. Early dropouts may be underrepresented, and some variables, like parental education, employment, student dedication, may be self-reported and incomplete. Engagement measures may also reflect infrastructure availability or device usage rather than actual participation. Finally, identifiers are anonymized, which may reduce the precision of longitudinal tracking, and the data does not include periods affected by the COVID-19 pandemic, meaning it may not capture disruptions caused by virtual or hybrid learning environments. Despite these limitations, the dataset provides a detailed framework for studying factors influencing student retention and academic success.


#### Import Dataset

We import the 2018 and 2022 subsets. This is because we will utilize the 2018 as the main body of data, and will then use the 2022 to determine a more accurate label. Th"is is described in the cleaning/preprocessing section.


In [47]:
# A - Load the Dataset (Just the 2022 and 2018 Subset)
data2_2022 = pd.read_csv(
    f'{RAW_DATA_DIR}university_dropout_2022.zip',
    sep=';',
    compression='zip' # Webpage download default as .zip
)

data2_2018= pd.read_csv(
    f'{RAW_DATA_DIR}university_dropout_2018.zip',
    sep=';',
    compression='zip' # Webpage download default as .zip
)


/tmp/ipykernel_420/1332526074.py:2: DtypeWarning: Columns (48,57,64,65,164) have mixed types. Specify dtype option on import or set low_memory=False.
  data2_2022 = pd.read_csv(
/tmp/ipykernel_420/1332526074.py:8: DtypeWarning: Columns (48,57,61,64,65) have mixed types. Specify dtype option on import or set low_memory=False.
  data2_2018= pd.read_csv(


#### Data Tidiness and Size

Based on the description of the dataset from the source on Zemodo, the dataset has been thouroughly tidied up, which is demontrated below. This will be shown for both the 2018 and the 2022 datasets seperately.

In [48]:
# B - Tidiness
def tidiness_function(df, title):
    print(f"--- {title} ---")
    
    # Show that each row is a single observation by cross checking duplicates against the identifiers for student, course, and degree hashes
    duplicates = df.duplicated(subset=['dni_hash', 'asi_hash', 'anyo_ingreso'])
    print("Number of duplicate rows:", duplicates.sum())
    
    # Show that columns are aptly named
    print('='*50)
    print(df.columns)
    
    # Show a preview of what the data looks like, demonstrating that columns are properly named, there are no overlapping values, and columns are generally meaningful
    print('='*50)
    return df.head(10)



In [49]:
tidiness_function(data2_2018, 'Tidiness of the Dataset from 2018')

--- Tidiness of the Dataset from 2018 ---
Number of duplicate rows: 0
Index(['dni_hash', 'tit_hash', 'asi_hash', 'anyo_ingreso', 'tipo_ingreso',
       'nota10_hash', 'nota14_hash', 'campus_hash', 'estudios_p_hash',
       'estudios_m_hash',
       ...
       'n_resource_days_2019_7', 'pft_events_2019_8', 'pft_days_logged_2019_8',
       'pft_visits_2019_8', 'pft_assignment_submissions_2019_8',
       'pft_test_submissions_2019_8', 'pft_total_minutes_2019_8',
       'n_wifi_days_2019_8', 'resource_events_2019_8',
       'n_resource_days_2019_8'],
      dtype='object', length=178)


,dni_hash,tit_hash,asi_hash,anyo_ingreso,tipo_ingreso,nota10_hash,nota14_hash,campus_hash,estudios_p_hash,estudios_m_hash,...,n_resource_days_2019_7,pft_events_2019_8,pft_days_logged_2019_8,pft_visits_2019_8,pft_assignment_submissions_2019_8,pft_test_submissions_2019_8,pft_total_minutes_2019_8,n_wifi_days_2019_8,resource_events_2019_8,n_resource_days_2019_8
0,922787046182,1d0ebea552eb,024e9489dcc6,"2013,0",NAP,"7,541","10,548",6781b441c78d2643,T,F,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,922787046182,1d0ebea552eb,007015a20672,"2013,0",NAP,"7,541","10,548",6781b441c78d2643,T,F,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,922787046182,1d0ebea552eb,0bac83d121af,"2013,0",NAP,"7,541","10,548",6781b441c78d2643,T,F,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,922787046182,1d0ebea552eb,88111a7f4d4f,"2013,0",NAP,"7,541","10,548",6781b441c78d2643,T,F,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,922787046182,1d0ebea552eb,dc1cdc6ee450,"2013,0",NAP,"7,541","10,548",6781b441c78d2643,T,F,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
5,922787046182,1d0ebea552eb,d1710e41e0ad,"2013,0",NAP,"7,541","10,548",6781b441c78d2643,T,F,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
6,922787046182,1d0ebea552eb,24a3dc9b4497,"2013,0",NAP,"7,541","10,548",6781b441c78d2643,T,F,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
7,922787046182,1d0ebea552eb,8f97bed7c6ca,"2013,0",NAP,"7,541","10,548",6781b441c78d2643,T,F,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
8,922787046182,1d0ebea552eb,e3913b66f36f,"2013,0",NAP,"7,541","10,548",6781b441c78d2643,T,F,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
9,922787046182,1d0ebea552eb,2b31968e912d,"2013,0",NAP,"7,541","10,548",6781b441c78d2643,T,F,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [50]:
tidiness_function(data2_2022, 'Tidiness of the Dataset from 2022')

--- Tidiness of the Dataset from 2022 ---
Number of duplicate rows: 0
Index(['dni_hash', 'tit_hash', 'asi_hash', 'anyo_ingreso', 'tipo_ingreso',
       'nota10_hash', 'nota14_hash', 'campus_hash', 'estudios_p_hash',
       'estudios_m_hash',
       ...
       'n_resource_days_2023_6', 'pft_events_2023_7', 'pft_days_logged_2023_7',
       'pft_visits_2023_7', 'pft_assignment_submissions_2023_7',
       'pft_test_submissions_2023_7', 'pft_total_minutes_2023_7',
       'n_wifi_days_2023_7', 'resource_events_2023_7',
       'n_resource_days_2023_7'],
      dtype='object', length=169)


,dni_hash,tit_hash,asi_hash,anyo_ingreso,tipo_ingreso,nota10_hash,nota14_hash,campus_hash,estudios_p_hash,estudios_m_hash,...,n_resource_days_2023_6,pft_events_2023_7,pft_days_logged_2023_7,pft_visits_2023_7,pft_assignment_submissions_2023_7,pft_test_submissions_2023_7,pft_total_minutes_2023_7,n_wifi_days_2023_7,resource_events_2023_7,n_resource_days_2023_7
0,319636fc9270,620c9c332101,4596fcf257c4,"2012,0",NAP,NaN,"9,456",e4f95d56d90df35e,F,L,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,319636fc9270,620c9c332101,81f4b5a1d0a8,"2012,0",NAP,NaN,"9,456",e4f95d56d90df35e,F,L,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,319636fc9270,620c9c332101,442fcac005ed,"2012,0",NAP,NaN,"9,456",e4f95d56d90df35e,F,L,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,319636fc9270,620c9c332101,3dc87ab71825,"2012,0",NAP,NaN,"9,456",e4f95d56d90df35e,F,L,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,319636fc9270,620c9c332101,677c622c0bfb,"2012,0",NAP,NaN,"9,456",e4f95d56d90df35e,F,L,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
5,319636fc9270,620c9c332101,2344965e8b89,"2012,0",NAP,NaN,"9,456",e4f95d56d90df35e,F,L,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
6,319636fc9270,620c9c332101,5f52e54c6a9c,"2012,0",NAP,NaN,"9,456",e4f95d56d90df35e,F,L,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
7,319636fc9270,620c9c332101,8b8b029f1142,"2012,0",NAP,NaN,"9,456",e4f95d56d90df35e,F,L,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
8,319636fc9270,620c9c332101,705d739be21c,"2012,0",NAP,NaN,"9,456",e4f95d56d90df35e,F,L,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
9,319636fc9270,620c9c332101,696d9363dc5a,"2012,0",NAP,NaN,"9,456",e4f95d56d90df35e,F,L,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [51]:
# C - Size of Dataset
print("Dataset from 2018")
print("Dataset shape (rows, columns):", data2_2018.shape)
print("Number of observations of student-course-year (rows):", data2_2018.shape[0])
print("Number of variables (columns):", data2_2018.shape[1])

print('='*50)
print("Dataset from 2022")
print("Dataset shape (rows, columns):", data2_2022.shape)
print("Number of observations of student-course-year (rows):", data2_2022.shape[0])
print("Number of variables (columns):", data2_2022.shape[1])

Dataset from 2018
Dataset shape (rows, columns): (152446, 178)
Number of observations of student-course-year (rows): 152446
Number of variables (columns): 178
Dataset from 2022
Dataset shape (rows, columns): (159173, 169)
Number of observations of student-course-year (rows): 159173
Number of variables (columns): 169


#### Data Missingness

As also mentioned in the paper connected to this dataset, there is a high systematic relationship in the missingness of much of the data, as well as a large portion of columns that have a lot of missing data. This is demonstrated below in the 2018 dataset, which can also be replicated in the 2022 dataset. The paper details some general trends in the missingness, which are most notable in the Wifi access records and digital record attributes due to different courses offerings taught under various instructors prefering different levels of LMS engagement. The 2018 dataset itself doesn't even have this data because it wasn't recorded in 2018 when it was collected, which is seen below. Another trend is shown and explained below.

In [52]:
# D - Missing Data Exploration
# Basic exploratory analysis on the missing data as porportions and counts
missing_counts = data2_2018.isnull().sum()
missing_percent = (missing_counts / len(data2_2018)) * 100

missing_df = pd.concat([missing_counts, missing_percent], axis=1)
missing_df.columns = ['missing_count', 'missing_pct']

missing_df = missing_df[missing_df['missing_count'] > 0].sort_values(by='missing_pct', ascending=False)
missing_df

,missing_count,missing_pct
pft_assignment_submissions_2018_10,152446,100.000000
n_wifi_days_2019_2,152446,100.000000
pft_days_logged_2018_12,152446,100.000000
pft_days_logged_2019_4,152446,100.000000
pft_visits_2019_4,152446,100.000000
...,...,...
anyo_ingreso,32699,21.449562
grupos_por_tipocredito_hash,30413,19.950015
nota_asig_hash,12649,8.297364
estudios_m_hash,906,0.594309


In [53]:
# D - Missing Data Exploration
# A deeper dive into why some data is missing in the way it is
# Let us take a look at the admission entrance grade, nota14_hash, and the tipo_ingresso, or the admission type
data2_2018.groupby('tipo_ingreso')['nota14_hash'] \
     .apply(lambda x: x.isnull().mean()*100) \
     .sort_values(ascending=False)

tipo_ingreso
NAI    88.771400
NCA    82.322357
NCF    46.923077
NAD    46.575342
NTE    39.465409
NIE     9.090909
NSC     8.747698
BMA     4.702194
ANT     1.030928
ASA     0.000000
NAP     0.000000
NLE     0.000000
NRO     0.000000
NSA     0.000000
NUE     0.000000
Name: nota14_hash, dtype: float64

This clearly demonstrates that some students come from backgrounds from which they are admitted to their university clearly aligns with the missingness of their entrance grade. For example, students who enter via vocational training or international transfers will not have these typical high school entrance exams. This is just one highlight of the systematic missingness of the data, but with 178 columns, some of which are highly related to one another, these relationships are common, especially in regards to missingness. 

#### Outlier Discovery

The accompanying paper describes that during the data anonymization of students, suspicious variables were dealt with to protect anonymity. For example, if a certain aggregation of variables could identify a student, this was deleted. Furthermore duplicate entries were deleted. Furthermore, a general check to ensure data types remained consistent and that value ranges for data was well within the expected distribution accross datasets was conducted. 

#### Data Cleaning/Preprocessing

This dataset features a high rate of missingness. A test (not shown) revealed that attempting a super agressive drop of all rows with any sort of missing data would cut the dataset to only a few hundred entries, so this is also not used. However, entries with all NA entries were deleted. Something to note in the paper is that some columns utilize NaN or N/A as a way to denote no or a specific class on a categorical variable. For example, one column is whether or not they are currently actively matriculated. In this instance, the number 1 is used to denote yes, and N/A is used to denote no. Another notable feature is that the csv was ';' deliminated and utilized commas as decimals, which is quite typical of much of Europe. As such, numbers are cleaned into decimal format and converted to float/int. 

Another important aspect is that the target label doesn't completely match the first dataset. The first dataset has an explicit target label with the categories being dropout, graduated, and currently enrolled. However, this dataset's target label is a binary between dropped out and not dropped out.  Furthermore, because this data only strictly provides a relative scale on individual course grades in the years of 2018, 2021, and 2022, we cannot calculate direct semester GPAs (we do not have the semester each course is taken in, nor do we have the number of units). 

Finally, this dataset is constructed such that each entry is a student-course enrollment entry. As such, even though there are around 150,000 entries in this dataset for 2018, there are only around 20,000 unique students, and even less are first year students. Thus, the original dataframe needs to be grouped by the student hash. 

A majority of columns are also dropped. Metrics like demographics, WiFi/LMS engagement, and socioeconomic factors, while probably impactful and could be helpful in predicting dropout, are not the specific focus for this research paper. As such, these columns are dropped. The code below goes into more specifics about what columns to keep, the label column construction, and general cleaning. 

In [80]:
# F - Data Cleaning
data_cleaned_2 = data2_2018.copy()

# Step 1: Drop rows that are all NA
data_cleaned_2 = data_cleaned_2.dropna(axis=0, how='all')

# Step 2: Keep only the columns related to academic performance that matches the previous dataset's dataframe, and rename them to maintain consistency. 
cols_to_keep = [
    # Organization/Bookkeeping Variables
    'dni_hash',
    'curso_mas_bajo',

    # Academic indicators
    'cred_sup_normal',
    'cred_sup_espec',
    'cred_mat_practicas',
    'cred_mat_sem_a',
    'cred_mat_sem_b',
    'cred_sup_sem_a',
    'cred_sup_sem_b',
    'nota_asig_hash',

    # Indicator
    'abandono_hash'
]
data_cleaned_2 = data_cleaned_2[cols_to_keep].copy()

# Step 3: Ensure numeric columns are numeric
numeric_cols = cols_to_keep.copy()
numeric_cols.remove('dni_hash')
numeric_cols.remove('abandono_hash')


# Step 3: Convert comma-based numbers to floats
for col in numeric_cols:
    data_cleaned_2[col] = (
        data_cleaned_2[col]
        .astype(str)                    
        .str.replace(',', '.')  
        .astype(float)                 
    )
    
# Step 4: Fill in any necessary N/A columns with values

# Step 5: Aggregate Indicator Variables, Filter to First Year Students
data_cleaned_2 = data_cleaned_2[
    data_cleaned_2['curso_mas_bajo'] == 1
]

data_cleaned_2 = (
    data_cleaned_2
    .groupby('dni_hash', as_index=False)
    .agg({
        'cred_sup_normal': 'first',
        'cred_sup_espec': 'first',
        'cred_mat_practicas': 'first',
        'cred_mat_sem_a': 'first',
        'cred_mat_sem_b': 'first',
        'cred_sup_sem_a': 'first',
        'cred_sup_sem_b': 'first',
        'nota_asig_hash': 'mean',
        'abandono_hash': 'first'
    })
)

data_cleaned_2.loc[
    (data_cleaned_2['cred_sup_sem_a'] == 0) &
    (data_cleaned_2['cred_sup_sem_b'] == 0),
    'nota_asig_hash'
] = 0

# Step 6: Construct target label by distinguishing "not dropout" to 'graduated' or 'still enrolled' after 4 years
data_2022_min['cred_pend_sup_tit'] = (
    data_2022_min['cred_pend_sup_tit']
    .astype(str)                        
    .str.replace(',', '.', regex=False) 
    .astype(float)                       
)

data_cleaned_2['target'] = data_cleaned_2['abandono_hash'].map({'A': 0, 'B': -1.0})

data_cleaned_2 = data_cleaned_2.merge(
    data_2022_min[['dni_hash', 'cred_pend_sup_tit']],
    on='dni_hash',
    how='left'
)

mask = data_cleaned_2['target'] == -1.0

data_cleaned_2.loc[mask, 'target'] = data_cleaned_2.loc[mask, 'cred_pend_sup_tit'].apply(
    lambda x: 2 if pd.isna(x) or x < 12 else 1
)

# Step 7: Rename columns
name_map_new = {
    'cred_sup_normal': 'Credits_Passed_Exam',
    'cred_sup_espec': 'Credits_Passed_Special',
    'cred_mat_practicas': 'Credits_Internship_Enrolled',
    'cred_mat_sem_a': 'Sem1_Enrolled',
    'cred_mat_sem_b': 'Sem2_Enrolled',
    'cred_sup_sem_a': 'Sem1_Passed',
    'cred_sup_sem_b': 'Sem2_Passed',
    'nota_asig_hash': 'GPA'
}
data_cleaned_2 = data_cleaned_2.rename(columns=name_map_new)

cols_to_drop = ['abandono_hash', 'cred_pend_sup_tit'] 
data_cleaned_2 = data_cleaned_2.drop(columns=[col for col in cols_to_drop if col in data_cleaned_2.columns])

data_cleaned_2.head()

,dni_hash,Credits_Passed_Exam,Credits_Passed_Special,Credits_Internship_Enrolled,Sem1_Enrolled,Sem2_Enrolled,Sem1_Passed,Sem2_Passed,GPA,target
0,0015a14d0e61,36.0,0.0,0.0,30.0,34.5,12.0,22.5,4.236364,1.0
1,0016fdba43a0,42.0,0.0,0.0,30.0,30.0,12.0,30.0,4.811111,2.0
2,001871270d17,60.0,0.0,0.0,9.0,12.0,9.0,12.0,6.787500,2.0
3,0018d4d04ecd,0.0,0.0,0.0,18.0,18.0,0.0,0.0,0.000000,2.0
4,003140d5d783,42.0,0.0,0.0,30.0,30.0,12.0,30.0,5.588889,2.0


In [81]:
print("Shape of cleaned dataframe:", data_cleaned_2.shape)

print("\nData types:")
print(data_cleaned_2.dtypes.value_counts())

processed_file_path = os.path.join(PROCESSED_DATA_DIR, 'university_dropout_cleaned_2022.csv')
data_cleaned_2.to_csv(processed_file_path, index=False, sep=';')

print(f"\nCleaned dataset saved to: {processed_file_path}")

Shape of cleaned dataframe: (7619, 10)

Data types:
float64    9
object     1
Name: count, dtype: int64

Cleaned dataset saved to: data/02-processed/university_dropout_cleaned_2022.csv


### Combining the Datasets

Even though the datasets are not completely aligned on all categories, for the most important ones in regards to our research question, they are. As such, we combine them into a single dataframe. 


## Results

### Exploratory Data Analysis

Instructions: replace the words in this subsection with whatever words you need to setup and preview the EDA you're going to do.   

Please explicitly load the fully wrangled data you will use from `data/02-processed`.  This is a good idea rather than forcing people to re-run the data getting / wrangling cells above.  Sometimes it takes a long time to get / wrangle data compared to reloading the fixed up dataset.

Carry out whatever EDA you need to for your project in the code cells below.  Because every project will be different we can't really give you much of a template at this point. But please make sure you describe the what and why in text here as well as providing interpretation of results and context.

Please note that you should consider the use of python modules in your work.  Any code which gets called repeatedly should be modularized. So if you run the same pre-processing, analysis or visualiazation on different subsets of the data, then you should turn that into a function or class.  Put that function or class in a .py file that lives in `modules/`.  Import the module you made and use it to get your work done.  For reference see `get_raw()` which is inside `modules/get_data.py`. 



#### Section 1 of EDA - please give it a better title than this

Some more words and stuff.  Remember notebooks work best if you interleave the code that generates a result with properly annotate figures and text that puts these results into context.

In [17]:
## YOUR CODE HERE
## FEEL FREE TO ADD MULTIPLE CELLS PER SECTION
data1_eda = pd.read_csv('data/02-processed/predict_students_dropout_clean.csv', sep=';')
data1_eda

,Previous qualification (grade),Admission grade,Gender,Sem1_Credited,Sem1_Enrolled,Sem1_Evaluations,Sem1_Passed,Sem1_Gpa,Sem1_Skipped,Sem2_Credited,Sem2_Enrolled,Sem2_Evaluations,Sem2_Passed,Sem2_Gpa,Sem2_Skipped,Target,Target_Numeric
0,122.0,127.3,1,0,0,0,0,0.000000,0,0,0,0,0,0.000000,0,Dropout,0
1,160.0,142.5,1,0,6,6,6,14.000000,0,0,6,6,6,13.666667,0,Graduate,2
2,122.0,124.8,1,0,6,0,0,0.000000,0,0,6,0,0,0.000000,0,Dropout,0
3,122.0,119.6,0,0,6,8,6,13.428571,0,0,6,10,5,12.400000,0,Graduate,2
4,100.0,141.5,0,0,6,9,5,12.333333,0,0,6,6,6,13.000000,0,Graduate,2
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
4419,125.0,122.2,1,0,6,7,5,13.600000,0,0,6,8,5,12.666667,0,Graduate,2
4420,120.0,119.0,0,0,6,6,6,12.000000,0,0,6,6,2,11.000000,0,Dropout,0
4421,154.0,149.5,0,0,7,8,7,14.912500,0,0,8,9,1,13.500000,0,Dropout,0
4422,180.0,153.8,0,0,5,5,5,13.800000,0,0,5,6,5,12.000000,0,Graduate,2


In [18]:
## Convert two GPA columsn to 4.0 scale for our understanding
## Also average GPAs over both semesters and add as a new column
max_port = 20.0
min_port = 0.0
max_usa = 4.0

grade_cols = [
    'Sem1_Gpa',
    'Sem2_Gpa'
]

for col in grade_cols:
    data1_eda[col] = (data1_eda[col] / max_port) * max_usa

data1_eda['Avg_Gpa'] = (data1_eda['Sem1_Gpa'] + data1_eda['Sem2_Gpa']) / 2

print(data1_eda[['Sem1_Gpa','Sem2_Gpa', 'Avg_Gpa']].head(100))

    Sem1_Gpa  Sem2_Gpa   Avg_Gpa
0   0.000000  0.000000  0.000000
1   2.800000  2.733333  2.766667
2   0.000000  0.000000  0.000000
3   2.685714  2.480000  2.582857
4   2.466667  2.600000  2.533333
..       ...       ...       ...
95  2.800000  2.650000  2.725000
96  2.733333  2.666667  2.700000
97  2.750000  2.895556  2.822778
98  2.840000  2.760000  2.800000
99  2.392000  0.000000  1.196000

[100 rows x 3 columns]


#### Section 2 of EDA if you need it  - please give it a better title than this

Some more words and stuff.  Remember notebooks work best if you interleave the code that generates a result with properly annotate figures and text that puts these results into context.

In [20]:
## YOUR CODE HERE
## FEEL FREE TO ADD MULTIPLE CELLS PER SECTION

## Ethics

Instructions: REPLACE the contents of this cell with your work, including any updates to recover points lost in your proposal feedback

## Team Expectations 

Instructions: REPLACE the contents of this cell with your work, including any updates to recover points lost in your proposal feedback

## Project Timeline Proposal

Instructions: Replace this with your timeline.  **PLEASE UPDATE your Timeline!** No battle plan survives contact with the enemy, so make sure we understand how your plans have changed.  Also if you have lost points on the previous checkpoint fix them